wave_optimization/tests/ceres/ba_test.cpp

```diff
@@ -0,0 +1,195 @@
+    EXPECT_FLOAT_EQ(0.0, residual1.fy);
+    EXPECT_FLOAT_EQ(0.0, residual1.cx);
+    EXPECT_FLOAT_EQ(0.0, residual1.cy);
+    EXPECT_FLOAT_EQ(0.0, residual1.x);
+    EXPECT_FLOAT_EQ(0.0, residual1.y);
+
+    // TEST CONSTRUCTOR WITH PARAMETERS
+    Mat3 K;
+    // clang-format off
+  K << 1.0, 0.0, 3.0,
+       0.0, 2.0, 4.0,
+       0.0, 0.0, 1.0;
+    // clang-format on
+    Vec2 x = Vec2{130, 62};
+
+    BAResidual residual2(K, x);
+    EXPECT_FLOAT_EQ(K(0, 0), residual2.fx);
+    EXPECT_FLOAT_EQ(K(1, 1), residual2.fy);
+    EXPECT_FLOAT_EQ(K(0, 2), residual2.cx);
+    EXPECT_FLOAT_EQ(K(1, 2), residual2.cy);
+    EXPECT_FLOAT_EQ(x(0), residual2.x);
+    EXPECT_FLOAT_EQ(x(1), residual2.y);
+}
+
+TEST(BAResidual, testResidual) {
+    // create vo dataset
+    VOTestDatasetGenerator generator;
+    generator.configure(TEST_CONFIG);
+    auto dataset = generator.generate();
+    LandmarkMap landmarks = dataset.landmarks;
+
+    // translate
+    auto true_G_p_GB = dataset.states[0].robot_G_p_GB;
+    auto t_vec = true_G_p_GB.data();
+
+    // orientation of robot Body in Global frame
+    const auto q_GB = dataset.states[0].robot_q_GB;
+
+    // rotate the body frame by this to get the camera frame
+    // or, use this to transform points from camera frame to body frame
+    Quaternion q_BC{Eigen::AngleAxisd(-M_PI_2, Vec3::UnitZ()) *
+                    Eigen::AngleAxisd(0, Vec3::UnitY()) *
+                    Eigen::AngleAxisd(-M_PI_2, Vec3::UnitX())};
+
+    Quaternion q_GC = q_GB * q_BC;
+
+    auto q_vec = q_GC.coeffs().data();
+
+    double e[2] = {0.0, 0.0};
+    const auto &feature = dataset.states[0].features_observed[0].second;
+    const auto &landmark_id = dataset.states[0].features_observed[0].first;
+
+    // test and assert
+    BAResidual residual{dataset.camera_K, feature};
+    residual(q_vec, t_vec, landmarks[landmark_id].data(), e);
+    EXPECT_NEAR(0.0, e[0], 0.0001);
+    EXPECT_NEAR(0.0, e[1], 0.0001);
+}
+
+TEST(BundleAdjustment, solve) {
+    // create vo dataset
+    VOTestDatasetGenerator generator;
+    generator.configure(TEST_CONFIG);
+    auto dataset = generator.generate();
+
+    // build bundle adjustment problem
+    BundleAdjustment ba;
+
+    // We'll keep the parameters in a vector for now.
+    // Note we have to pre-allocate the vector, so that pointers to the elements
+    // don't change as we insert.
+    std::vector<Vec3> params_G_p_GC(dataset.states.size());
+    std::vector<Quaternion> params_q_GC(dataset.states.size());
+    LandmarkMap params_landmarks = dataset.landmarks;
+
+    // Add offset to landmark initial estimates
+    for (auto &l : params_landmarks) {
+        l.second += Vec3{0.3, -0.3, 0.3};
+    }
+
+    // Add pose parameters
+    for (size_t i = 0; i < dataset.states.size(); i++) {
+        // translation
+        const auto &true_G_p_GC = dataset.states[i].robot_G_p_GB;
+        // add offset
+        Vec3 G_p_GC = true_G_p_GC + Vec3{0.5, 0.1, -0.5};
```

Comment:
Yes, maybe it should be gaussian.. then again this is the initial estimate, not a synthetic measurement, so neither is exactly realistic. Realistically the estimate would come from IMU or odometry, but this example doesn't have them.